certstore_linux.go

```diff
@@ -1,14 +1,336 @@
 package certstore
 
-import "errors"
+/*
+#cgo pkg-config: nss
+#include <nss.h>
+#include <pk11pub.h>
+#include <nspr.h>
+#include <prio.h>
+#include <prerror.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <pwd.h>
+#include <p12.h>
+#include <p12plcy.h>
+#include <stdlib.h>
+#include <cryptohi.h>
 
-// This will hopefully give a compiler error that will hint at the fact that
-// this package isn't designed to work on Linux.
-func init() {
-	CERTSTORE_DOESNT_WORK_ON_LINIX
+SECItem *P12U_NicknameCollisionCallback(SECItem *old_nick, PRBool *cancel, void *wincx) {
+	char *nick = NULL;
+	SECItem *ret_nick = NULL;
+	CERTCertificate *cert = (CERTCertificate *)wincx;
+	if (!cancel || !cert) {
+		fprintf(stdout, "cancel: %p, cert: %p\n", cancel, cert);
+		return NULL;
+	}
+	nick = CERT_MakeCANickname(cert);
+	if (!nick) {
+		fprintf(stdout, "nick %p\n", nick);
+		return NULL;
+	}
+	if (old_nick && old_nick->data && old_nick->len &&
+		PORT_Strlen(nick) == old_nick->len &&
+		!PORT_Strncmp((char *)old_nick->data, nick, old_nick->len)) {
+		PORT_Free(nick);
+		fprintf(stdout, "old_nick %p, nick %p\n", old_nick, nick);
+		return NULL;
+    }
+	//fprintf(stdout, "using nickname: %s\n", nick);
+	ret_nick = PORT_ZNew(SECItem);
+	if (ret_nick == NULL) {
+		PORT_Free(nick);
+		fprintf(stdout, "ret_nick %p\n", ret_nick);
+		return NULL;
+	}
+	ret_nick->data = (unsigned char *)nick;
+	ret_nick->len = PORT_Strlen(nick);
+	return ret_nick;
+}
+
+CERTCertListNode *CertListHead(CERTCertList *l) {
+	return CERT_LIST_HEAD(l);
+}
+
+CERTCertListNode *CertListNext(CERTCertListNode *n) {
+	return CERT_LIST_NEXT(n);
+}
+
+int CertListEnd(CERTCertListNode *n, CERTCertList *l) {
+	return CERT_LIST_END(n, l);
+}
+*/
+import "C"
+import (
+	"crypto"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/user"
+	"math/big"
+	"path"
+	"unicode/utf16"
+	"unsafe"
+)
+
+type nssStore int
+
+func (nssStore) Identities() ([]Identity, error) {
+	var identities = make([]Identity, 0)
+	//fmt.Printf("Listing certificates:\n")
+	var certs = C.PK11_ListCerts(C.PK11CertListAll, nil)
+	if certs == nil {
+		C.NSS_Shutdown()
+		return nil, fmt.Errorf("Error %d, closing and returing...\n", int(C.PR_GetError()))
+	}
+	var node *C.CERTCertListNode
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		identity := nssIdentity(*node)
+		identities = append(identities, &identity)
+	}
+	return identities, nil
+}
+
+type nssIdentity C.CERTCertListNode
+
+func (i *nssIdentity) Signer() (crypto.Signer, error) {
+	return i, nil
+}
+
+func (i *nssIdentity) Certificate() (*x509.Certificate, error) {
+	var der = i.cert.derCert
+	var bytes = C.GoBytes(unsafe.Pointer(der.data), C.int(der.len))
+	cert, err := x509.ParseCertificate(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
```

Comment:
> I'm not reviewing this method, since it will hopefully be replaced with a call to `NSSCertificate_BuildChain`.

I used `CERT_GetCertChainFromCert` instead - it was simpler to use and seems to do the right thing...